graphene_parser.py

```diff
@@ -449,185 +279,77 @@ def _parse_generic_function_impl(
 
-        for generic_token in generic_definitions_tree.children:
-            assert isinstance(generic_token, Token)
-            if generic_token.name == "GENERIC_IDENTIFIER":
-                generic = parse_generic_definition(generic_token)
-                generic_definitions.append(generic)
-
-                if generic.is_value_arg:
-                    generic_mapping[generic.name] = cg.GenericValueReference(
-                        generic.name
-                    )
-                else:
-                    generic_mapping[generic.name] = cg.UnresolvedGenericType(
-                        generic.name
-                    )
-            else:
-                assert generic_token.name == "GENERIC_PACK_IDENTIFIER"
+        for generic in node.generic_definitions:
+            if generic.is_packed:
                 # NOTE enforced by the grammar. We might want to relax this
                 # later and allow more than one pack per function.
                 assert variadic_type_pack_name is None
-                variadic_type_pack_name = generic_token.value
-
-        unresolved_return = TypeTransformer.parse(return_type_tree, generic_mapping)
-
-        arg_names: list[str] = []
-        unresolved_args: list[cg.UnresolvedType] = []
-        for name, type_tree in in_pairs(args_tree.children):
-            assert isinstance(name, Token)
-            # TODO refactor, this is very ugly.
-            if isinstance(type_tree, Token):
-                # TODO user-facing error.
-                assert type_tree.value == variadic_type_pack_name
-                variadic_args_pack_name = f"{name.value}..."
+                variadic_type_pack_name = generic.name
+                continue
+
+            generic_definitions.append(parse_generic_definition(generic))
+            if isinstance(generic, lp.TypeGenericDefinition):
+                generic_mapping[generic.name] = cg.UnresolvedGenericType(generic.name)
             else:
-                assert isinstance(type_tree, Tree)
-                arg_names.append(name.value)
+                assert isinstance(generic, lp.NumericGenericDefinition)
+                generic_mapping[generic.name] = cg.GenericValueReference(generic.name)
 
-                unresolved_args.append(
-                    TypeTransformer.parse(type_tree, generic_mapping)
-                )
+        parser = ParseType(generic_mapping)
+        specialization = [
+            parser.parse_specialization(item) for item in node.specialization
+        ]
+        unresolved_return = parser.parse_type(node.return_)
+
+        # TODO: this is a temporary hack
+        if len(node.args) > 0 and isinstance(node.args[-1][1], lp.PackType):
+            variadic_args_pack_name, packed_type = node.args.pop()
+
+            # For now a packed type must be a plain type name
+            assert isinstance(packed_type, lp.PackType)
+            assert isinstance(packed_type.type_, lp.NamedType)
```

Comment:
So a `PackType` is itself a `Type`, but also contains another `Type`, whilst a `Type` is just a `Meta`?

Reply:
This is the curse and the brilliance of oop :cry: 

- I'm using `Type` as an interface here: anything that can go where you'd normally put a `Type` is a type.
- `Type` like everything the lexer parser produces inherits from `ParsedNode` (which only contains `meta`).
- `PackType` conforms to the `Type` interface because it's used as a type in the function signature. It also contains another `Type` (left of the '...'), atm the only type we support here is a `NamedType` with no specialization hence `NamedType...`. Maybe we want to add references too? Or maybe specializations :upside_down_face: as a way to implement concepts :upside_down_face:?

Reply:
Ok that makes sense I guess